chore: Update Meta Graph API versions

rgon opened the pull request:
Fixes https://linear.app/chatwoot/issue/PLA-7/update-the-api-version-to-v24
## Description

The v13 API used in the cloud inbox was deprecated and set as end-of-life @ 28 of may 2024. According to FB's website, it shouldn't be available. Source: https://developers.facebook.com/docs/graph-api/changelog/

This PR simply updates its version to v20, to avoid service interruption.

In the future, it may be wise to update the version as mentioned in:
https://github.com/chatwoot/chatwoot/blob/23e30fcb1a0d8492fb93b53716b4a4c48ad00903/app/services/whatsapp/providers/whatsapp_cloud_service.rb#L66
But since it appears that the API should be unusable soon, I'm pushing this change.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

I patched my installation with these changes and ran it. It successfully receives messages, sends messages, loads templates.


## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

rgon committed:
chore: bump Facebook API to v20 in Whatsapp Cloud service to avoid using v13 deprecated API

pranavrajs commented:
@rgon Thanks for the changes. Have you verified if there are any significant API payload changes in the transition from v13 to v20? I believe this should be done before merging the PR.

rgon commented:
@pranavrajs Yes, in two ways.

I am currently using it in production, with no side effects. Subjectively, since I have updated the webhook API version in the Meta Developer Console, it subjectively triggers faster (have no way of measuring it, but the developer console alerts that delays could happen with mismatched versions). 

Also, analyzing the entire changelog, NO, there are no applicable changes. Below a review of the entire API changelog:

----

Reading the Whatsapp API changelog is a tough task. Even though it's accessed though the versioned FB API, the changelog is not groped by versions, but rather dates... See for v14, the only newer version that mentions Whatsapp: https://developers.facebook.com/docs/graph-api/changelog/version14.0#whatsapp where it links to a separate changelog: https://developers.facebook.com/docs/whatsapp/business-platform/changelog

Here, I have compiled a list with all the API Changes. It's a shame that the automatic versioning tool does not work for the v13 API https://developers.facebook.com/tools/api_versioning

Since v13 was released in 2022/02/08 and v14 was released in 2022/05/25, there are the following changes within that timeframe in the Whatsapp API (Filtered for the Cloud API).

+ [2022/03/01: Support for video message types](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#video-messages)
+ [2022/03/15: List messages, reply buttons, preview URL, documentation changes](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#list-reply-preview)
+ [2022/04/05: messages with stickers, deregister phone number API, new phone verification option, new referral webhooks](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#cloud-api-apr-5-2022)
+ [2022/05/03: fixed disappearing messages, better thumbnail quality](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#cloud-api-may-3-2022)
+ [2022/05/19: public cloud API Release](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#cloud-api-may-19-2022)
+ [2022/05/19: v14.0 message template categories deprecated template types for v14.0](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#bm-api-may-19-2022) ** - not applicable

And after those, here is every breaking change (where a version number is mentioned)

+ [2022/11/22: v15.0 New Webhooks Fields and Values, New Business Endpoint Filters, New Phone Number Node Fields](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#cloud-api-may-19-2022) However, no fields removed
+ [2023/02/02: ](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#february-2--2023)
    + Changes on Templates created or edited - N/A
    + Error subcodes, which are rarely used and should not be relied upon, will no longer be included in v16.0+ error responses. Use code and details instead. - Not used
    + Error code 100 ("code":100) will now use Invalid parameter as their title - N/A
    + errors object in webhooks now include message and error_data.details properties - N/A
 + [2023/04/04: v16+ Category validation is now part of the template review process](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#april-4--2023)
+ [2023/05/23](https://developers.facebook.com/docs/whatsapp/business-platform/changelog#april-4--2023): v17+ 
    + Attempting to send a paused template in a template message using v17 or later now returns [error code 132015](https://developers.facebook.com/docs/whatsapp/cloud-api/support/error-codes#other-errors) instead of 132001.
    + Attempting to send a disabled template in a template message using v17 or later now returns [error code 132016](https://developers.facebook.com/docs/whatsapp/cloud-api/support/error-codes#other-errors) instead of 132001.

This last 2023/05/23 v17 change is the most likely to affect. However, chatwoot does not check for error codes, but rather prints them out:
https://github.com/chatwoot/chatwoot/blob/ce8e1ec93dbf2c0d5253da82bf4c0cdd6f16dc66/app/services/whatsapp/send_on_whatsapp_service.rb#L8C7-L8C20

facuxt commented:
this means that every user that create a new facebook app (now it only lets you select api v20) it will not work the integration with cloud api <-> chatwoot?

sojan-official committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

sojan-official commented:
@muhsin-k  could you test these changes ?

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

abielzulio commented:
can we move this quickly? i think it is causing us to not be able to get the whatsapp messages correctly as we only have 2 api versions from meta: v22 and v23.

https://github.com/chatwoot/chatwoot/issues/11725
https://github.com/chatwoot/chatwoot/issues/11725

vrijmetse commented:
We need this fix, please get the ball rolling.

vrijmetse commented:
@sojan-official what is going on here ?

muhsin-k commented:
Thanks for raising this PR. We will test these changes and will merge the PR if there are no issues with the integration.

@rgon Could you please fix the conflicts?

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
chore: fix specs

muhsin-k committed:
chore: update version

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

sojan-official commented:
## Pending Tasks 
-  We will need to fix the tests and test the overall changes 

## Cases to confirm
- Whether this would have any impact on older customer who were under the older api version ? ( ref: https://github.com/chatwoot/chatwoot/pull/9811#issuecomment-2258309365 - things should be okay )

----


@muhsin-k  will take this forward and get this merged  ( target date : 12 Aug )

cc: @sojan-official, @tds-1

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

sojan-official committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
chore: update to v24

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation

muhsin-k committed:
Merge branch 'develop' into deps-upgrade-facebook-api-version-avoids-deprecation
